compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeResolver.java

```diff
@@ -1459,7 +1459,7 @@ private void fillEffectiveType(BIntersectionType intersectionType,
         while (iterator.hasNext()) {
             BType type = iterator.next();
             BLangType bLangType = bLangTypeItr.next();
-            if (type.tag == TypeTags.READONLY) {
+            if (Types.getImpliedType(type).tag == TypeTags.READONLY) {
```

Comment:
Shall we extract this into a separate variable? When generating the effective type, we once again obtain the implied type within the `calculateEffectiveType` function.

Reply:
updated in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/42025/commits/61e343e1616c463b2edaa6953ef0df4b986250d4).